### PR TITLE
RavenDB-12137 [WIP] Add graph queries to RunningQueries

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 1]
```diff
@@ -1436,6 +1445,18 @@ public long GetEnvironmentsHash()
 
             return hash;
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AddToCurrentlyRunningGraphQueries(ExecutingQueryInfo executingQueryInfo)
+        {
+            CurrentlyRunningGraphQueries.Add(executingQueryInfo);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Can we move it to GraphQueryRunner and acces it via Database.QueryRunner.Graph.CurrentlyRunningGraphQueries ?

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 2]
```diff
@@ -392,7 +393,15 @@ public void NotifyOnPendingClusterTransaction()
         private long? _nextClusterCommand;
         private long _lastCompletedClusterTransaction;
         public long LastCompletedClusterTransaction => _lastCompletedClusterTransaction;
+        private int _numberOfGraphQueries;
+
+        public int GetNextGraphQueryId()
+        {
+            return Interlocked.Increment(ref _numberOfGraphQueries);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Can we move it to GraphQueryRunner and acces it via Database.QueryRunner.Graph.CurrentlyRunningGraphQueries ?

File: src/Raven.Server/Documents/Handlers/Debugging/QueriesDebugHandler.cs [hunk 3]
```diff
@@ -17,16 +18,23 @@ public class QueriesDebugHandler : DatabaseRequestHandler
         [RavenAction("/databases/*/debug/queries/kill", "POST", AuthorizationStatus.ValidUser)]
         public Task KillQuery()
         {
+            ExecutingQueryInfo query;
             var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("indexName");
             var id = GetLongQueryString("id");
+            if (name == "@graph")
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Debugging/QueriesDebugHandler.cs]: '@graph' should be moved to constants e.g. Constants.Documents.Indexing.GraphIndexName

File: src/Raven.Server/Documents/Handlers/Debugging/QueriesDebugHandler.cs [hunk 4]
```diff
@@ -68,29 +98,32 @@ public Task RunningQueries()
...
-                        writer.WritePropertyName((nameof(query.QueryId)));
-                        writer.WriteInteger(query.QueryId);
-                        writer.WriteComma();
+                if(indexes.Count != 0)
+                    writer.WriteComma();
 
-                        writer.WritePropertyName((nameof(query.StartTime)));
-                        writer.WriteDateTime(query.StartTime, isUtc: true);
-                        writer.WriteComma();
+                writer.WritePropertyName("Graph queries");
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Debugging/QueriesDebugHandler.cs]: Can we use property without space? Maybe `@graph` that you used before?

File: src/Raven.Server/Documents/Queries/GraphQueryDoneRunning.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+
+namespace Raven.Server.Documents.Queries
+{
+    public struct GraphQueryDoneRunning : IDisposable
+    {
+        private static readonly TimeSpan DefaultLockTimeout = TimeSpan.FromSeconds(3);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Queries/GraphQueryDoneRunning.cs]: Is this used?

File: src/Raven.Server/Documents/Queries/GraphQueryDoneRunning.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+        private static readonly TimeSpan ExtendedLockTimeout = TimeSpan.FromSeconds(30);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Queries/GraphQueryDoneRunning.cs]: Is this used?

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: Yes, this shouldn't be on `DocumentDatabase`

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 7]
```diff
@@ -392,7 +393,15 @@ public void NotifyOnPendingClusterTransaction()
...
+        }
+
         private int _clusterTransactionDelayOnFailure = 1000;
+        public readonly ConcurrentSet<ExecutingQueryInfo> CurrentlyRunningGraphQueries = new ConcurrentSet<ExecutingQueryInfo>();
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Why do we need a separate behavior for graph queries? Can't we use the existing running queries for the same thing?

File: src/Raven.Server/Documents/Handlers/Debugging/QueriesDebugHandler.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Debugging/QueriesDebugHandler.cs]: I really don't like it being separate. 
These are just queries, and should be treated the same.

**Review by ayende — CHANGES_REQUESTED** [review]
I don't like this impl.
What we want to do here is to use the exact same code and process that we have for the already existing running queries impl, as well as the same cancellation support.

That means that we call the regular `AddToRunningQueries` and (maybe) handle any changes there, but I think we probably shouldn't have to do much there.

**Comment by ppekrol** [comment]
@ayende this is because running queries were only on index-level

@talweiss1982 I think we need to take this opportunity and move running queries to QueryRunner, so we can handle all types of queries (index, collection, graph). Thoughts?

**Comment by ayende** [comment]
fine by me

**Comment by maximburyak** [comment]
@ayende @ppekrol can we postpone the change proposed be @ppekrol to 4.2.x? I think that right now we want to concentrate on getting all graph features done

**Comment by ppekrol** [comment]
We will do it in 2 stages.
1. @talweiss1982 will move the code to GraphQueryRunner (as was written in one of the first comments) and add the cancellation support
2. To allow @talweiss1982 to continue the work on Graph issues, I will do the unification

**Comment by talweiss1982** [comment]
@ppekrol i added cancelation to graph queries and moved the running graph queries under the runner

**Review by ppekrol — APPROVED** [review]
Rebased and did some minor adjustments.

**Review by ayende — APPROVED** [review]

**Pull request merged**